### PR TITLE
Kinetis uVisor stack guard

### PR DESCRIPTION
**Pull request by orenc17A**

### CODE REVIEW

File: core/vmpu/inc/vmpu_kinetis_mem.h [hunk 1]
```diff
@@ -31,7 +31,24 @@
...
  */
-#define UVISOR_TACL_BACKGROUND 0x000007D0U
+
+
+ #define MPU_BUS_MASTER_CORE 0
+ #define MPU_BUS_MASTER_DEBUGGER 1
+ #define MPU_RGDn_WORD2_Mx_UM_NONE_SM_RW     0x00000010U
+ #define MPU_RGDn_WORD2_Mx_UM_NONE_SM_NONE   0x00000018U
+ #define MPU_RGDn_WORD2_Mx_UM_RWX_SM_RWX     0x00000007U
+ #define MPU_RGDn_WORD2_Mx(bus_master, perm) (perm << (bus_master * 6))
```
Comment (alzix) [inline core/vmpu/inc/vmpu_kinetis_mem.h]: only valid for 4 frist bus master

File: core/vmpu/inc/vmpu_kinetis_mem.h [hunk 2]
```diff
@@ -31,7 +31,24 @@
...
+
+#define UVISOR_TACL_DEBUGGER_BACKGROUND ( \
+    MPU_RGDn_WORD2_Mx(MPU_BUS_MASTER_CORE, MPU_RGDn_WORD2_Mx_UM_NONE_SM_NONE) | \
+    MPU_RGDn_WORD2_Mx(MPU_BUS_MASTER_DEBUGGER, MPU_RGDn_WORD2_Mx_UM_RWX_SM_RWX) \
```
Comment (alzix) [inline core/vmpu/inc/vmpu_kinetis_mem.h]: debugger access it allowed by background 0 region - please remove

File: core/linker/default.h [hunk 3]
```diff
@@ -197,10 +197,17 @@ SECTIONS
         __uvisor_stack_start_boundary__ = .;
         . += STACK_GUARD_BAND;
 #endif
+#ifdef ARCH_MPU_KINETIS
+    . = ALIGN(32);
```
Comment (danny4478) [inline core/linker/default.h]: Add comment why 32 bytes alignment is needed or use an informative define like KINETIS_REGION_ALIGNMENT

File: core/vmpu/inc/vmpu_kinetis_mem.h [hunk 4]
```diff
@@ -31,7 +31,24 @@
...
+)
+
+#define UVISOR_TACL_CORE_BACKGROUND ( \
+    MPU_RGDn_WORD2_Mx(MPU_BUS_MASTER_CORE, MPU_RGDn_WORD2_Mx_UM_NONE_SM_RW) | \
+    MPU_RGDn_WORD2_Mx(MPU_BUS_MASTER_DEBUGGER, MPU_RGDn_WORD2_Mx_UM_RWX_SM_RWX) \
```
Comment (danny4478) [inline core/vmpu/inc/vmpu_kinetis_mem.h]: I am probably missing something here: Is 1 shifted 42 bits left?

File: core/vmpu/src/mpu_kinetis/vmpu_kinetis.c [hunk 5]
```diff
@@ -56,6 +56,10 @@ uint32_t vmpu_sys_mux_handler(uint32_t lr, uint32_t msp)
     uint32_t pc;
     uint32_t fault_addr, fault_status;
 
+    // this is an example code that causes hardfault by touching the stack guard
+    // uint32_t *guard_region = ((uint32_t*)&__uvisor_stack_start_boundary__) + 4;
+    // *guard_region = 1;
+
```
Comment (danny4478) [inline core/vmpu/src/mpu_kinetis/vmpu_kinetis.c]: Good.
Just don't forget to remove.

File: core/vmpu/src/mpu_kinetis/vmpu_kinetis_mpu.c [hunk 6]
```diff
@@ -38,18 +38,20 @@
...
+ * |    6    | <-- Box Stack, K64F_MPU_REGIONS_STATIC
  * +---------+
- * |    2    | <-- Application SRAM unlock
- * |    1    | <-- Application Flash unlock
- * |    0    | <-- Background region
+ * |    5    | <-- Application SRAM unlock
+ * |    4    | <-- Application Flash unlock
+ * |    3    | <-- Background region (stack start to 0xFFFFFFFF)
+ * |    2    | <-- Background region (0 to stack guard)
+ * |    1    | <-- Background region (for debugger)
```
Comment (danny4478) [inline core/vmpu/src/mpu_kinetis/vmpu_kinetis_mpu.c]: Was region 0 never used?

File: core/vmpu/src/mpu_kinetis/vmpu_kinetis_mpu.c [hunk 7]
```diff
@@ -28,7 +28,7 @@
 #include "configurations.h"
 
 /* The K64F has 12 MPU regions, however, we use region 0 as the background
- * region with `UVISOR_TACL_BACKGROUND` as permissions.
+ * region with `UVISOR_TACL_DEBUGGER_BACKGROUND` as permissions.
  * Region 1 and 2 are used to unlock Application SRAM and Flash.
```
Comment (danny4478) [inline core/vmpu/src/mpu_kinetis/vmpu_kinetis_mpu.c]: This is not true anymore for regions 1 and 2, right?

File: core/linker/default.h (hunk 3 above)
Comment (orenc17A) [reply core/linker/default.h]: comment added

File: core/vmpu/src/mpu_kinetis/vmpu_kinetis_mpu.c (hunk 6 above)
Comment (orenc17A) [reply core/vmpu/src/mpu_kinetis/vmpu_kinetis_mpu.c]: graphics corrected

File: core/vmpu/inc/vmpu_kinetis_mem.h (hunk 1 above)
Comment (orenc17A) [reply core/vmpu/inc/vmpu_kinetis_mem.h]: added a comment

File: core/vmpu/inc/vmpu_kinetis_mem.h (hunk 2 above)
Comment (orenc17A) [reply core/vmpu/inc/vmpu_kinetis_mem.h]: fixed

File: core/vmpu/src/mpu_kinetis/vmpu_kinetis_mpu.c (hunk 7 above)
Comment (orenc17A) [reply core/vmpu/src/mpu_kinetis/vmpu_kinetis_mpu.c]: fixed docs

**Comment by orenc17A** [comment]
@alzix can we ends this?

**Pull request merged**